### ISSUE TITLE
[Issue/Question] .cfignore does not respect folders

### ISSUE THREAD

**Issue by jd-carroll** [cloudfoundry/cli#564]
Using the latest from mean.js, I installed the latest cli tools and I get the following:

```
$ cf push
Using manifest file c:\dev\artecha\inventory-management-meanjs\manifest.yml

Updating app mean in org inventory-management / space development as jdsalingerjr@gmail.com...
OK

Creating route mean-unsaccharine-arboriculture.cfapps.io...
OK

Binding mean-unsaccharine-arboriculture.cfapps.io to mean...
OK

Uploading mean...
FAILED
Error uploading application.
GetFileAttributesEx c:\dev\artecha\inventory-management-meanjs\node_modules\bower\node_modules\updat
e-notifier\node_modules\latest-version\node_modules\package-json\node_modules\got\node_modules\duple
xify\node_modules\end-of-stream\node_modules\once\node_modules\wrappy\package.json: The system canno
t find the path specified.
```

The `node_modules` is very clearly listed as an entry in the `.cfignore` file as `node_modules/`.

Is this a bug or am I missing something?

**Comment by simonleung8** [comment]
The `node_modules/`will not be uploaded, however, CLI walks through all directories in order to determine what files need to be uploaded. Are you running into this issue under Windows? IIRC, Windows has a limit on how long a dir name can be.

**Comment by KrekkieD** [comment]
I explicitly set `node_modules` to be ignored in `.cfignore` and still it iterates through this dir. This is a really annoying thing on windows which indeed occurs once paths become too long, which depends on the modules you have installed.

I think it should give the `.cfignore` paths more respect before attempting iterating over something that serves no purpose.

**Comment by jd-carroll** [comment]
@simonleung8 That's completely ridiculous... Why would it care what's in an ignored directory? Can you please point me to where this code is and I will submit a PR to fix this issue.

**Comment by petereberlein** [comment]
I ran into the same problem that the CLI cannot deal with long paths on Windows, however, in my use case the folder structure is intended, ie. there is no .cfignore in it because I need all the files in this way.

It is certainly a good idea to not traverse folders with a .cfignore in it and that solves the issue for the original reporter, but I'd appreciate it very much, if the actual problem (the use of ancient Windows APIs that cannot deal with long paths) was fixed as well.

**Comment by culmor30** [comment]
Same issue here; `node_modules/` is listed in my .cfignore file but my app fails to push unless I move that folder elsewhere, due to cf looking through the deeply-nested directory tree. This wasn't happening with an older version of CF CLI (just upgraded today).

**Comment by dkoper** [comment]
Thanks for the additional details. I have prioritized the story in our backlog.

**Comment by srankers** [comment]
Same issue on Win7 for cf-cli-installer_6.13.0_winx64. The downgrade to  v6.12.0 didn't work for me.
(checked both versions 32 an 64 Bit)
>cf --version
cf version 6.12.0-8c65bbd-2015-06-29T17:18:54+00:00
>cf --version
cf version 6.12.0-8c65bbd-2015-06-30T00:16:56+00:00

**Comment by maral** [comment]
What about this bug, are you guys dealing with this? Pretty annoying...

**Comment by krishicks** [comment]
We're looking at this issue now.

**Comment by krishicks** [comment]
Hi,

The latest edge release has a potential fix for this issue. I say potential because the story hasn't yet passed acceptance, but the binary is available at https://github.com/cloudfoundry/cli#downloads.

Please give it a try if you're experiencing this problem and let me know if it works.

**Comment by PeteAshworth** [comment]
just wanted to say the update worked for me.  Thank you!

**Comment by culmor30** [comment]
The version you linked fixed it for me, thanks!

**Comment by krishicks** [comment]
Does the other project have node_modules in its .cfignore?

**Comment by krishicks** [comment]
Oh, weird. The whole message didn't show up on mobile until I submitted the response.

**Comment by PeteAshworth** [comment]
Worked for me; glad to get this bug out of the way. Thanks
```
C:\>cf --version
cf version 6.14.0+f12b776-2015-12-15
```

**Comment by dkoper** [comment]
@culmor30 Can you please open a separate issue for your remaining issue?
Please include reproducible steps for us.

**Comment by culmor30** [comment]
Looks like I got an older build than @PeteAshworth somehow. I'll try the 12-25 build later today and post here if it's working, otherwise I'll open a new issue.

**Comment by krishicks** [comment]
@culmor30 You'll need to download the edge binary from https://github.com/cloudfoundry/cli#downloads

**Comment by culmor30** [comment]
Yup, my mistake. Edge binary fixed the problem, thanks very much!

**Comment by dkoper** [comment]
This fix is included in cf CLI release v6.14.1 published today.
https://github.com/cloudfoundry/cli#downloads

**Comment by isreekan** [comment]
I'm still hitting this issue trying with version 6.15 . FAILED
Error processing app files in 'C:\Data\2016\git\reveal.js-3.2.0\reveal.js-3.2.0': GetFileAttributesEx C:\Data\2016\git\reveal.js-3.2.0\revea
l.js-3.2.0\node_modules\grunt-contrib-cssmin\node_modules\maxmin\node_modules\pretty-bytes\node_modules\meow\node_modules\normalize-package-
data\node_modules\is-builtin-module\node_modules\builtin-modules\builtin-modules.json: The system cannot find the path specified.

C:\Data\2016\git\reveal.js-3.2.0\reveal.js-3.2.0>cf --version
cf version 6.15.0+fa1bfe2-2016-01-12

**Comment by dkoper** [comment]
Hi @isreekan,

You may be hitting #728, of which the fix will be included in 6.16.0. You can try an Edge build and if it doesn't fix it, please open a new issue with reproducible steps.

Cheers,
Dies Koper
CF CLI PM

**Comment by isreekan** [comment]
@dkoper thanks . The edge build worked fine. The other workaround was to add a .cfignore for the nested modules

**Comment by el-davo** [comment]
I have the latest cf installed and still when i put node_modules entry in .cfignore it still tries to upload all node_modules files. Why is this issue closed?

**Comment by dkoper** [comment]
@el-davo, it was closed because several people reported cf CLI 6.16.0 resolved the issue.
Maybe you've encountered a new issue showing the same behavior?
Please submit a new issue with details and we'll take a look!

Cheers,
Dies Koper
CF CLI PM
